### ISSUE TITLE
Use Kotlin default values instead of force-feeding null when mapping POJOs

### ISSUE THREAD

**Issue by axelfontaine** [jOOQ/jOOQ#10236]
### Expected behavior and actual behavior:

When jOOQ maps a record onto the following Kotlin POJO:

```kotlin
data class Race(
  val id: Long,
  val distance: Int = 0
)
```

if the distance column is null, or if the resultset doesn't include a distance column, jOOQ currently forces a null down the throat of the data class' primary constructor, resulting in an NPE at runtime.

Detecting whether a constructor is non-nullable with a default value and using that instead would be much nicer.

### Versions:

- jOOQ: 3.13.2

**Comment by axelfontaine** [comment]
The current workaround for this is to change the data class as follows:

```kotlin
data class Race(
  val id: Long,
  private val _distance: Int?
) {
    val distance: Int; get() = if (_distance != null) _distance else 0
}
```

**Comment by lukaseder** [comment]
Thank you very much for your suggestion. There are a few things to distinguish here:

> if the distance column is null

I think that a `null` value in the result set *must* be represented by a `null` value in client code, unless there is an explicit `Converter` or `Binding` that does something else. In SQL, `NULL` and `DEFAULT` are different concepts. For example, an `UpdatableRecord.store()` operation will expose `DEFAULT` behaviour if the column value was not set explicitly, but `NULL` behaviour if it was set to `null` explicitly.

So I don't think this should be fixed.

Side note: On the other hand, `Convert` already converts a `null` value to `0` if the target type is `int.class` (as opposed to `Integer.class`. The contributed logic from #7324 might need to take this into account, not sure. Does Kotlin not produce an `int` type for `Int` in byte code? However, this would just be hiding the actual problem. It wouldn't work for `String` types.

> or if the resultset doesn't include a distance column

In that case, indeed, the `distance` property should not be set on a Kotlin data class, as we're using a named parameter approach as per #7324.

In this case, since we're not supposed to set the parameter, the default value should indeed apply, independently of the type is nullable or not!

We're currently reusing logic calling named parameters in Java (in case those are available to the reflection API). The Java logic initialises all arguments to `null`, which is probably producing this case.

**Comment by lukaseder** [comment]
> * Pojo field is nullable and has a default value -> ??

Well, here, the suggested behaviour is clearly wrong! Consider this vanilla kotlin code, with no jOOQ involvement:

```kotlin
fun main() {
    println(C())
    println(C(null))
}

data class C(val i: Int? = 1)
```

The output is:

```
C(i=1)
C(i=null)
```

So, an explicit `null` is an acceptable value that must not be overridden, and it is different from an absent value (just like SQL's `NULL` and `DEFAULT` expressions are different things)! Additionally, such a change would be drastic, backwards incompatible change of behaviour, which has very little chance of being considered in jOOQ.

I understand that the intention of the first case can be "Do The Right Thing&trade;". I'm still reluctant to do it. The value *is* `null`, maybe for a reason (that may have been overlooked), so applying the default *may* not be the intention of the programmer, and removing the error will be against a "fail fast" philosophy that might apply here.

```kotlin
fun main() {
    println(C())
    println(C(null!!))
}

data class C(val i: Int = 1)
```

Resulting in:

```
C(i=1)
Exception in thread "main" java.lang.NullPointerException
 at FileKt.main (File.kt:7) 
 at FileKt.main (File.kt:-1) 
 at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (:-2) 
```

Can we just say this `null` is safe to ignore? I really don't know.  We're only hearing the opinions of the folks preferring the change, not of the folks liking the status quo, so there may just as well not be any problem here.

Note, that field nullability would force you to think about this, explicitly, and would probably make the status quo more acceptable?

- https://github.com/jOOQ/jOOQ/issues/13999

At least in the case of type safe mappers, not the reflective ones.

**Comment by lukaseder** [comment]
Having said so, you can implement a `RecordMapperProvider` to override the default behaviour to your liking:
https://www.jooq.org/doc/latest/manual/sql-execution/fetching/pojos-with-recordmapper-provider/

**Comment by gofabian** [comment]
I understand your point. Maybe it is better to not do it.


> Having said so, you can implement a `RecordMapperProvider` to override the default behaviour to your liking: https://www.jooq.org/doc/latest/manual/sql-execution/fetching/pojos-with-recordmapper-provider/

Yes, a workaround could be something like this:

```kotlin
class JacksonRecordMapperProvider : RecordMapperProvider {

    private val objectMapper = jacksonObjectMapper()
        .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
        .setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)

    override fun <R : Record, E : Any?> provide(recordType: RecordType<R>, type: Class<out E>): RecordMapper<R, E> {
        return RecordMapper<R, E> { record ->
            val map = record.intoMap()
            objectMapper.convertValue(map, type)
        }
    }

}
```

**Comment by lukaseder** [comment]
> I understand your point. Maybe it is better to not do it.

Yeah. There's always a tradeoff. This is definitely a nuisance in some cases, but it's important that the *actual problem* is properly understood, and that we don't jump to conclusions about what the best solution should be. In a way, this is an instance of the https://xyproblem.info

Let's look again at the original case:

> ```kotlin
> data class Race(
>   val id: Long,
>   val distance: Int = 0
> )
> ```

What's the value of using `0` instead of `null` for races without known distance? A Race can't really have a distance of `0` in most sports / competitions / etc., so this `0` acts as a substitute for `null`, where `null` is a much more appropriate value.

Sometimes, values like `distance` are aggregate values from a SQL `SUM(distance)` over race sections, for example. SQL has a peculiar way of making aggregate functions over empty sets return `NULL` even when the aggregate function's underlying operation (`+`) has an obvious identity value, that of `0`. This can be annoying in other paradigms that don't follow SQL, but here we are. `SUM(distance) IS NULL` means that the set was empty (including e.g. due to a `LEFT JOIN`).

Can the `DefaultRecordMapper` really be that bold and just assume this `NULL` value isn't semantic for at least some users?

Again, I understand the nuisance, but I'm just not convinced that this is really a mapping problem (it may be, but may also not be)
